### ISSUE TITLE
market location for point of consumption

### ISSUE THREAD

**Issue by dottorer** [buzzn/scrum#160]
**User story**

As Pho
I want clearly track the point of consumption (fka "Entnahmestelle")
so that I can handle all cases of changes during a billing cycle.

---

**Definition of done**

- [ ] a PT contract has one `market_location` and one `register`.
- [ ] a market location has an 11-digit `market_location_id` ("MaLo-Id") which conforms with the formatting rules.
- [ ] a market location has one or more meters with their respective one or more registers.
- [ ] a market location has a name (like "Wohnung 15, 1.OG rechts")
- [ ] a register still is the **metering location** and has the 33-char metering point id.

--

**XD**

- ...

---
**Technical details**

_Frontend_

- [ ] ...

_Backend_

- [ ] ...

---
**Information**
- [Wiki on Marktlokation](https://de.wikipedia.org/wiki/Marktlokations-Identifikationsnummer)
- [Info on ML](http://www.natuvion.com/marktlokation/)

**Comment by dottorer** [comment]
@Justus3000 @evaklopp @phillipoertel @mkristian After consultation with Justus, I am putting forth this story to (re-)introduce the point of consumption ("Entnahmestelle") in the correct form of "Marktlokation". It's the law since Feb1 and we need it for the billing bricks conceptually anyways.

Please review this story and its DoD and correct where wrong. I am a bit shaky on the details I must admit....

**Comment by dottorer** [comment]
Input from @Justus3000 (and StefanE):

Alignment mit Legalist/Metering Specialist SE hat folgendes ergeben:
1. MaLo and MeLo sind Begriffe der Energiewirtschaft und sollten nur im Zusammenhang mit Drittversorgten, ÜGZ und Produktionszähler(n) verwendet werden.
2. Für’s UI und alle “Zählpunkte”, die nicht auf dem Radarschirm des Imperiums auftauchen (sprich alle LSN-Zählpunkte) sollten wir auf diese Begriffe verzichten; die linke Spalte in Evas Tabelle sollte “Bezugsstelle” heißen, ein Begriff, den wir schon lange und in diversen anderen Kontexten erfolgreich verwenden.
3. Konzeptionell ist eine Bezugsstelle natürlich eine “MaLo”.    

Will work this into the story proposal.

**Comment by eharit** [comment]
XD: Mockups are ready to review

**Comment by evaklopp** [comment]
I quickly discussed the screens with Pali, but more questions appeared, so we need to discuss it with you @dottorer again. But I think we are very close to a solid concept, so much more clarity in it now!

**Comment by scsirdx** [comment]
@mkristian List of endpoints with registers:
1. `${apiUrl}${apiPath}/localpools/${groupId}/contracts/${contractId}?include=register:[meter],contrator_bank_account,contractor:[address],customer_bank_account,customer:[address,contact:address],tariffs,payments`
2. `${apiUrl}${apiPath}/localpools/${groupId}/power-taker-contracts?include=register,customer:[address,contact:address]`
3. `${apiUrl}${apiPath}/localpools/${groupId}/meters/${meterId}?include=registers,formula_parts:register`
4. `${apiUrl}${apiPath}/localpools/${groupId}/meters?include=registers:[readings,contracts:[customer]],formula_parts:register`
5. Add/delete reading are using format /meterId/registers/registerId
6. register, ticker, register PATCH, get readings
7. `${apiUrl}${apiPath}/persons/${userId}?include=address,contracts:[localpool,register]`

**Comment by mkristian** [comment]
@scsirdx 
following will remain as they are:
3. ${apiUrl}${apiPath}/localpools/${groupId}/meters/${meterId}?include=registers,formula_parts:register
4. ${apiUrl}${apiPath}/localpools/${groupId}/meters?include=registers:[readings,contracts:[customer]],formula_parts:register
5. Add/delete reading are using format /meterId/registers/registerId
6. register, ticker, register PATCH, get readings

all others are basically contracts and here the register relation will be replaced by market_location. still work in progress. but for the time being they are still intact.

**Comment by mkristian** [comment]
@scsirdx with 
4. ${apiUrl}${apiPath}/localpools/${groupId}/meters?include=registers:[readings,contracts:[customer]],formula_parts:register
do you need this nested contracts here ?

**Comment by mkristian** [comment]
@evaklopp @dottorer regarding the system overview and since @scsirdx asked my where this status is (https://projects.invisionapp.com/d/#/console/12205665/279020738/preview):

there was the idea to make connected discovergy meters/registers green, unconnected red and standard-profile ones grey. all three values are immutable over time, so IMO it is not a status (red might be an exception and can change to green but once it is green it will remain green).

the green: meter sent at least one reading to discovergy and then it is also visible on discovergy.com
the red: meter never ever sent a reading to discovergy and it will not even show on discovergy.com but we can see it in the api.discovergy.com (this might change as it could be just a bug that we can see it there and not on discovergy.com)

currently we are declaring the red discovergy meters/registers with datasource discovergy which is wrong as all those meters (from beekeeper) are not connected to the internet (or missing a piece of hardware for sending data) and needs to read manually and the 'datasource' should reflect this, i.e. they are as dumb as any standard-profile meters.

in short: the status is wrong, the datasource as we do it right now is wrong and should be an attribute picked by the operator (manual, discovergy).

and what we actually want here is
* red : meter is offline, i.e. last reading is older than 5 minutes
* green : meter is online i.e. last reading is not older than 5 minutes
but this requires some more work to achieve this in a performant way for list of meter/registers of group.
